packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/presentation/effects/SetBadgeCountNotificationEffect.java

```diff
@@ -0,0 +1,37 @@
+package expo.modules.notifications.notifications.presentation.effects;
+
+import android.app.Notification;
+import android.content.Context;
+
+import androidx.annotation.Nullable;
+import expo.modules.notifications.notifications.presentation.builders.BadgeSettingNotificationBuilder;
+import me.leolin.shortcutbadger.ShortcutBadger;
+
+public class SetBadgeCountNotificationEffect extends BaseNotificationEffect {
+  private static final String EXTRAS_BADGE_KEY = BadgeSettingNotificationBuilder.EXTRAS_BADGE_KEY;
+
+  public SetBadgeCountNotificationEffect(Context context) {
+    super(context);
+  }
+
+  @Override
+  public boolean onNotificationPresented(@Nullable String tag, int id, Notification notification) {
+    return applyBadgeFromNotification(notification);
+  }
+
+  @Override
+  public boolean onNotificationPresentationFailed(@Nullable String tag, int id, Notification notification) {
+    // We could also just return false here. Then, notifications that failed to be presented
+    // wouldn't affect badge count. Applying badge count from failed notifications lets us
+    // properly handle badge-update-only notifications.
```

Comment:
this seems a little funky to me. Are "badge-update-only" notifications a thing? 🤔 If so, it seems like they shouldn't get the "Failed" callback if there was intentionally nothing to present.

Reply:
Badge-only notifications are properly supported on iOS, so I was thinking—let's to do the same on Android. On iOS you can even send an only-badge-and-sound notification! (This would be implementable on Android too by playing a default notification sound in a side effect.)

I agree this is not an ideal solution, I was going back and forth between different ways to do this.

We could decide upfront whether to try to send the notification to the OS, so eg. `NotificationBuilder` could return something like `(sendToTheOS: boolean, notification)` and side effects could have three callbacks: `presented`, `failedToPresent`, `decidedNotToPresent` (maybe with some `enum` reason). This seems to me like a good next-stop solution.

We could add a fallback `title = request.title || ""` to the builder so that notification presenting would always succeed. Then we could add another side effect that would cancel the badge-only notifications immediately after display (notification without content looks bad). However, this solution seems even more itchy to me.

We could also explicitly not support badge-only notifications (as we do with badge-sound-only notifications), but… 😕

Reply:
OK. Reflecting on this some more, I think this makes sense and seems like the cleanest approach. Thanks for the clear and thoughtful explanations. And I suppose that even if a notification fails to be presented for some other reason, one would want the badge count to be updated anyway.